### PR TITLE
WIP: Win subsystems, cross build settings

### PR DESCRIPTION
**Pull request by lasote**
#2120 #2098 
- More friendly interface to run in windows subsystems
- Several tools.
- Also moved some tests and refactored other, to be used from the **conan-io/test** suite.
- #2171 should be merged first and remove from here the temporary patch of build require order
- Future docs: https://github.com/conan-io/docs/issues/436

### CODE REVIEW

File: conans/test/util/tools_test.py [hunk 1]
```diff
@@ -463,4 +465,8 @@ def download_retries_test(self):
 
         # Authorized using headers
         tools.download("https://httpbin.org/basic-auth/user/passwd", dest,
-                       headers={"Authorization": "Basic dXNlcjpwYXNzd2Q="}, overwrite=True)
\ No newline at end of file
+                       headers={"Authorization": "Basic dXNlcjpwYXNzd2Q="}, overwrite=True)
+
+
+    def cross_building_test(self):
```
Comment (memsharded) [inline conans/test/util/tools_test.py]: Remove this?

File: conans/test/integration/diamond_test.py [hunk 2]
```diff
@@ -20,8 +20,26 @@ def setUp(self):
...
+        self.diamond_tester.test(use_cmake=True)
+
+    def diamond_cmake_targets_test(self):
+        self.diamond_tester.test(use_cmake=True, cmake_targets=True)
+
+    def diamond_default_test(self):
+        self.diamond_tester.test(use_cmake=False)
+
+
+class DiamondTester(object):
```
Comment (memsharded) [inline conans/test/integration/diamond_test.py]: I don't see the advantage of the DiamondTester

File: conans/client/tools/oss.py [hunk 3]
```diff
@@ -221,10 +224,55 @@ def get_solaris_version_name(version):
...
             return "Solaris 11"
 
+    @staticmethod
+    def bash_path():
+        if "CONAN_BASH_PATH" not in os.environ:
+            bash_path = None
+            for ext in [".bat", ".exe", ".sh", ""]:
+                bash_path = bash_path or which("bash%s" % ext)
+        else:
+            bash_path = os.getenv("CONAN_BASH_PATH")
```
Comment (memsharded) [inline conans/client/tools/oss.py]: I prefer other flow, first doing os.getenv, if True => return it

File: conans/client/tools/oss.py [hunk 4]
```diff
@@ -221,10 +224,55 @@ def get_solaris_version_name(version):
         elif version.minor() == "5.11":
             return "Solaris 11"
 
+    @staticmethod
+    def bash_path():
+        if "CONAN_BASH_PATH" not in os.environ:
+            bash_path = None
+            for ext in [".bat", ".exe", ".sh", ""]:
```
Comment (memsharded) [inline conans/client/tools/oss.py]: So, will keep checking? Why not break?

File: conans/test/integration/basic_build_test.py [hunk 5]
```diff
@@ -50,16 +25,30 @@ def build_default_test(self):
...
-            logger.error("This platform does not support G++ command")
-            return
-        install = "install -s compiler=gcc -s compiler.libcxx=libstdc++ -s compiler.version=4.9"
-        for cmd, lang, static, pure_c in [(install, 0, True, True),
-                                          (install + " -o language=1 -o static=False", 1, False, False)]:
-            self._build(cmd, static, pure_c, use_cmake=False, lang=lang)
+            build(self, cmd, static, pure_c, use_cmake=False, lang=lang)
+
+
+def build(tester, cmd, static, pure_c, use_cmake, lang):
```
Comment (memsharded) [inline conans/test/integration/basic_build_test.py]: Why taking this outside of the class?

File: conans/test/utils/cpp_test_files.py [hunk 6]
```diff
@@ -34,7 +34,7 @@ def build(self):
 
                 command = ('{} && cl /EHsc main.cpp hello{}.lib {}'.format(vcvars, self.name, flags))
                 self.run(command)
-        elif self.settings.compiler == "gcc" and self.settings.os == "Linux":
+        elif tools.os_info.bash_path() and (tools.which("aclocal") or tools.which("aclocal.exe")):
```
Comment (memsharded) [inline conans/test/utils/cpp_test_files.py]: SHouldn't the ``tools.which`` found it even if you don't add the .exe?

File: conans/client/tools/win.py [hunk 7]
```diff
@@ -199,30 +201,63 @@ def escape_windows_cmd(command):
...
+        curdir = unix_path(cwd or os.getcwd(), path_flavor=subsystem)
         to_run = 'cd "%s"%s && %s ' % (curdir, hack_path, bashcmd)
-        custom_bash_path = os.getenv("CONAN_BASH_PATH", "bash")
-        wincmd = '%s --login -c %s' % (custom_bash_path, escape_windows_cmd(to_run))
+        wincmd = '%s --login -c %s' % (os_info.bash_path(), escape_windows_cmd(to_run))
         conanfile.output.info('run_in_windows_bash: %s' % wincmd)
-        conanfile.run(wincmd)
+        return conanfile.run(wincmd, win_bash=False)
+
+def vcvars_dict(*args, **kwargs):
```
Comment (memsharded) [inline conans/client/tools/win.py]: Is this method intended for end-users or only for the context-manager?

File: conans/test/integration/basic_build_test.py (hunk 5 above)
Comment (lasote) [reply conans/test/integration/basic_build_test.py]: It's used in the `conan-io/test` to build with mingw

File: conans/test/integration/diamond_test.py (hunk 2 above)
Comment (lasote) [reply conans/test/integration/diamond_test.py]: It's used in the `conan-io/test` to build a diamond with mingw

File: conans/client/tools/win.py (hunk 7 above)
Comment (lasote) [reply conans/client/tools/win.py]: I use it in my vs wrapper.

File: conans/test/utils/cpp_test_files.py (hunk 6 above)
Comment (lasote) [reply conans/test/utils/cpp_test_files.py]: I searched for the original PR, I thought there was some better reason to keep it like this, but probably yes, in windows should auto-try the extensions in the same order Windows acts.

https://github.com/conan-io/conan/pull/1801

Comment (lasote) [reply conans/test/utils/cpp_test_files.py]: Will take a look

File: conans/client/tools/oss.py (hunk 3 above)
Comment (lasote) [reply conans/client/tools/oss.py]: ok. good

File: conans/client/tools/oss.py (hunk 4 above)
Comment (lasote) [reply conans/client/tools/oss.py]: I think I tried to create kind of priority doing that. I'll check.

**Pull request merged**